contracts/nexus-gateway/src/contract/execute.rs

```diff
@@ -1,17 +1,54 @@
 use connection_router::state::Message;
-use cosmwasm_std::Response;
-use error_stack::Result;
+use cosmwasm_std::{Addr, Response};
+use error_stack::ResultExt;
 
 use crate::error::ContractError;
 use crate::nexus;
+use crate::state::Store;
 
-pub fn route_to_nexus(msgs: Vec<Message>) -> Result<Response<nexus::Message>, ContractError> {
-    let msgs = msgs
-        .into_iter()
-        .map(TryInto::try_into)
-        .collect::<Result<Vec<nexus::Message>, _>>()?;
+use super::Contract;
 
-    Ok(Response::new().add_messages(msgs))
+type Result<T> = error_stack::Result<T, ContractError>;
+
+impl<S> Contract<S>
+where
+    S: Store,
+{
+    pub fn route_messages(
+        mut self,
+        sender: Addr,
+        msgs: Vec<Message>,
+    ) -> Result<Response<nexus::Message>> {
+        match sender {
+            sender if sender == self.config.nexus => todo!(),
+            sender if sender == self.config.router => self
+                .route_to_nexus(msgs)
+                .change_context(ContractError::RouteToNexus),
+            _ => Err(ContractError::Unauthorized)?,
+        }
+    }
+
+    fn route_to_nexus(&mut self, msgs: Vec<Message>) -> Result<Response<nexus::Message>> {
+        let msgs = msgs
+            .into_iter()
+            .filter_map(|msg| match self.store.is_message_routed(&msg.cc_id) {
+                Ok(true) => None,
+                Ok(false) => Some(Ok(msg)),
+                Err(err) => Some(Err(err)),
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        for msg in &msgs {
+            self.store.set_message_routed(&msg.cc_id)?;
+        }
+
+        let msgs = msgs
+            .into_iter()
+            .map(TryInto::try_into)
+            .collect::<Result<Vec<nexus::Message>>>()?;
+
+        Ok(Response::new().add_messages(msgs))
```

Comment:
This message will definitely error if the nexus fails to route the message right? Just want to make sure a message isn't set as routed here, but then not actually routed in the nexus.

Reply:
yes

Reply:
This actually makes me wonder, why don't we track the messages in nexus, instead of here? Since nexus has the restriction of only once routing, why not just enforce that there?

Reply:
good point!

Reply:
This contract is an extension of nexus as the name suggests. The code here is more upgradable than nexus.